### PR TITLE
Feature email client drafts

### PR DESCRIPTION
**Pull request by daniel-samson**
Task: SCRM-555

## Description

This changes adds the drafts feature to the email client. When editing  an existing draft, the user is taken to a detail view for drafts. Which allows the user to manage the relationships to said email. The compose view doesn't (yet) have auto population for attachments. So for now we can use this new detail view as a stepping stone towards that. Since we don't have the "from" field in the compose view, I have opted to use the default email account to store the drafts.

**For Developers:**
The detail drafts view has it's own metadata (detaildraftviewdefs.php). This is so we can control action menu's, fields, etc. Plus it allows developers to extend the drafts in an upgrade safe way.

I have moved some of the field handling into a method so we can prevent code duplication. 

Update the setup method in the EditView, the DetailView, and the views so that you can pass the file name for the metadata file. This means that developers can now easily add additional metadafiles with minimal code duplication:

- Create metadata file
- set the key of the viewdefs to the name of the view

Example
`$viewdefs['Emails']['MyCustomeView']`

- In the view override the preDisplay method like so 
**Example:**
```
        $metadataFile = $this->getMetaDataFile();
        $this->ev = $this->getEditView();
        $this->ev->ss =& $this->ss;
        $this->ev->view = 'MyCustomeView';
        $this->ev->ss->assign('INBOUND_ID', $inboundEmailID);
        $this->ev->ss->assign('TEMP_ID', create_guid());
        $this->ev->ss->assign('RETURN_MODULE', isset($_REQUEST['return_module']) ? $_REQUEST['return_module'] : '');
        $this->ev->ss->assign('RETURN_ACTION', isset($_REQUEST['return_action']) ? $_REQUEST['return_action'] : '');
        $this->ev->setup(
            $this->module,
            $this->bean,
            $metadataFile,
            get_custom_file_if_exists('modules/Emails/include/ComposeView/ComposeView.tpl'
            true,
            'mycustomviewdefs');
```
- Or you can create your own CustomView and in your CustomView override the setup set the defaults
**Example**
```
    public function setup(
        $module,
        $focus  = null,
        $metadataFile = null,
        $tpl = 'include/DetailView/DetailView.tpl',
        $createFocus = true,
        $metadataFileName = 'editviewdefs'
    )
    {
        parent::setup($module, $focus, $metadataFile, $tpl, $createFocus, $metadataFileName);
    }

```
I have updated the PHP Doc for the EditView, DetailVIew. So IDE with indexing can use it to navigate through the code base better.

**Note:**
Going forward we will need to remove code duplication for in the Detail view and other specialisations. We should change name and path of the EditView because the class name and the filename do not match, which will allow future auto loading capabilities.

I also think we need a progress jquery plugin to so that we correctly handle the progress bar, and also so we can remove the code duplication.

## Motivation and Context

* Add feature back in

## How To Test This

* Repair and Rebuld
* Repair minified js
* Repair js groupings
* Create draft
*  Open list view (emails)
* Select folder
* Select draft folder for default account
* Click on the draft
* Select the options for the action menu and or manage the related subpanels

When edit drafting a draft
* Select the "Edit draft" on the action menu
* Send email

Test filters as there may be some issues with the filters.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] "New" feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: modules/Emails/views/view.detaildraft.php [hunk 1]
```diff
@@ -37,37 +39,37 @@
...
+    /**
+     * @see SugarView::preDisplay()
+     */
+    public function preDisplay()
+    {
+        $metadataFile = parent::getMetaDataFile();
+        $this->dv = new DetailDraftView();
+        $this->dv->populateBean();
+        $this->dv->ss =& $this->ss;
+        $this->dv->setup($this->module, $this->dv->focus, $metadataFile,
```
Comment (gymad) [inline modules/Emails/views/view.detaildraft.php]: Field 'focus' is not a member of class DetailDraftView (see above)

File: modules/Emails/views/view.savedraftemail.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ * display the words  'Powered by SugarCRM' and 'Supercharged by SuiteCRM'.
+ */
+
+if (!defined('sugarEntry') || !sugarEntry) {
+    die ('Not A Valid Entry Point');
+}
+
+class EmailsViewSavedraftemail extends ViewAjax
+{
+    public function __construct()
```
Comment (gymad) [inline modules/Emails/views/view.savedraftemail.php]: Missing php doc comment

File: modules/Emails/views/view.savedraftemail.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        $this->options['show_title'] = false;
```
Comment (gymad) [inline modules/Emails/views/view.savedraftemail.php]: Missing parent::__construct() call, also code duplication: This constructor do same as the parent?

File: modules/Emails/views/view.savedraftemail.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        $this->options['show_subpanels'] = false;
+        $this->options['show_search'] = false;
+    }
+
+    public function display()
+    {
+        global $app_strings;
+        $response = array();
+
+        if(empty($this->bean->status)) {
```
Comment (gymad) [inline modules/Emails/views/view.savedraftemail.php]: Can't found class property 'status'. Where is the definition for it?

File: modules/Emails/views/view.sendemail.php [hunk 5]
```diff
@@ -57,22 +57,40 @@ public function __construct()
 
     public function display()
     {
-        $response_code = 500;
+        global $app_strings;
+        $response = array();
 
         if(empty($this->bean->status)) {
```
Comment (gymad) [inline modules/Emails/views/view.sendemail.php]: Where the 'status' again?

File: modules/Emails/views/view.savedraftemail.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        $this->options['show_title'] = false;
+        $this->options['show_header'] = false;
+        $this->options['show_footer'] = false;
+        $this->options['show_javascript'] = false;
+        $this->options['show_subpanels'] = false;
+        $this->options['show_search'] = false;
+    }
+
+    public function display()
```
Comment (gymad) [inline modules/Emails/views/view.savedraftemail.php]: Function seems a code duplication. Similar function in view.sendemail.php, only one different a extra case in a switch. Can you refactoring this, please?

File: modules/Emails/Email.php [hunk 7]
```diff
@@ -3383,4 +3383,208 @@ public function revertFieldNullable($field)
...
             }
     	}
+
+    /**
+     * Uses the $_REQUEST to populate the fields of an Email SugarBean.
+     *
+     * @param Email $bean
+     * @return bool|Email|SugarBean
+     */
+    public function populateBeanFromRequest(Email $bean) {
```
Comment (daniel-samson) [inline modules/Emails/Email.php]: Renamed signature and passed $_REQUEST into method instead. It unfortunately takes 200 lines to populate the request. When refactoring after we have autoloading we should consider using PSR-7.

File: modules/Emails/Email.php [hunk 8]
```diff
@@ -3383,4 +3383,208 @@ public function revertFieldNullable($field)
...
+            }
+        }
+
+
+        $toEmailAddresses = preg_split('/[,;]/', $bean->to_addrs, null, PREG_SPLIT_NO_EMPTY);
+        $bean->to_addr_arr = array();
+        foreach ($toEmailAddresses as $ea => $address) {
+
+            $email = '';
+            $display = '';
```
Comment (daniel-samson) [inline modules/Emails/Email.php]: email and display are used see line 3507. these are temp variables used to work strip out the name from the email address (Joe Blogs <Joe@example.com>

File: modules/Emails/include/ComposeView/EmailsComposeView.js [hunk 9]
```diff
@@ -784,21 +813,200 @@
...
+          mb.showHeader();
+          mb.setBody(response.data.title);
+          mb.showFooter();
+          $(self).trigger("saveEmailSuccess", [self, response]);
+
+          if ($(self).closest('[name=ComposeView]').find('[name=id]').length === 0) {
+            var id = $('<input>').attr('type', 'hidden').attr('name', 'id').val(response.data.id);
+            id.appendTo($(self).closest('[name=ComposeView]'));
+          } else {
+            var id = $(self).closest('[name=ComposeView]').find('[name=id]');
```
Comment (daniel-samson) [inline modules/Emails/include/ComposeView/EmailsComposeView.js]: fixed at the time or writing this in an earlier commit - I wasn't completely sure where the compose would sit in the DOM or how we would define it.

File: modules/Emails/include/ComposeView/EmailsComposeView.js [hunk 10]
```diff
@@ -784,21 +813,200 @@
...
+        response = JSON.parse(response);
+        mb.setBody(response.errors.title);
+        $(self).trigger("saveEmailError", [self, response]);
+      }).always(function (response) {
+        response = JSON.parse(response);
+        $(self).trigger("saveEmailAlways", [self, response]);
+      });
+
+      return false;
+    }
```
Comment (daniel-samson) [inline modules/Emails/include/ComposeView/EmailsComposeView.js]: fixed

File: modules/Emails/views/view.savedraftemail.php (hunk 2 above)
Comment (daniel-samson) [inline modules/Emails/views/view.savedraftemail.php]: added phpdoc

File: modules/Emails/views/view.detaildraft.php (hunk 1 above)
Comment (daniel-samson) [inline modules/Emails/views/view.detaildraft.php]: added phpdoc

File: modules/Emails/include/ListView/ListViewDataEmails.php [hunk 11]
```diff
@@ -193,9 +207,9 @@ public function getListViewData(
 
             // Create a list of fields to filter and decide based on the field which type of filter to carry out
             $emailServerFilter = array();
```
Comment (daniel-samson) [inline modules/Emails/include/ListView/ListViewDataEmails.php]: it is used look at the loop below

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            require("modules/$this->module/metadata/detaildraftviewdefs.php");
+        }
+
+        $this->defs = $viewdefs[$this->module][$this->view];
+    }
+
+    public function populateBean()
+    {
+        if (!empty($_REQUEST['record'])) {
+            $this->focus = BeanFactory::getBean('Emails')->retrieve($_REQUEST['record']);
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+                global $app_strings;
+                $error = str_replace("[file]", "modules/$this->module/metadata/detaildraftviewdefs.php", $app_strings['ERR_CANNOT_CREATE_METADATA_FILE']);
+                $GLOBALS['log']->fatal($error);
+                echo $error;
+                die();
+            }
+            require("modules/$this->module/metadata/detaildraftviewdefs.php");
+        }
+
+        $this->defs = $viewdefs[$this->module][$this->view];
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: it comes from the parent. added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        //Check if inline editing is enabled for detail view.
+        if(!isset($sugar_config['enable_line_editing_detail']) || $sugar_config['enable_line_editing_detail']){
+            $this->ss->assign('inline_edit', true);
+        }
+        $this->focus = $focus;
+        $this->tpl = $tpl;
+        $this->module = $module;
+        $this->metadataFile = $metadataFile;
+        if(isset($GLOBALS['sugar_config']['disable_vcr'])) {
+            $this->showVCRControl = !$GLOBALS['sugar_config']['disable_vcr'];
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        $this->th->ss = $this->ss;
+
+        //Check if inline editing is enabled for detail view.
+        if(!isset($sugar_config['enable_line_editing_detail']) || $sugar_config['enable_line_editing_detail']){
+            $this->ss->assign('inline_edit', true);
+        }
+        $this->focus = $focus;
+        $this->tpl = $tpl;
+        $this->module = $module;
+        $this->metadataFile = $metadataFile;
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+        $this->th = new TemplateHandler();
+        $this->th->ss = $this->ss;
+
+        //Check if inline editing is enabled for detail view.
+        if(!isset($sugar_config['enable_line_editing_detail']) || $sugar_config['enable_line_editing_detail']){
+            $this->ss->assign('inline_edit', true);
+        }
+        $this->focus = $focus;
+        $this->tpl = $tpl;
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        global $sugar_config;
+
+        $this->th = new TemplateHandler();
+        $this->th->ss = $this->ss;
+
+        //Check if inline editing is enabled for detail view.
+        if(!isset($sugar_config['enable_line_editing_detail']) || $sugar_config['enable_line_editing_detail']){
+            $this->ss->assign('inline_edit', true);
+        }
+        $this->focus = $focus;
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        $focus  = null,
+        $metadataFile = null,
+        $tpl = 'include/DetailView/DetailView.tpl',
+        $createFocus = true
+    )
+    {
+        global $sugar_config;
+
+        $this->th = new TemplateHandler();
+        $this->th->ss = $this->ss;
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/include/DetailDraftView/DetailDraftView.php [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+if (!defined('sugarEntry') || !sugarEntry) {
+    die('Not A Valid Entry Point');
+}
+
+require_once('include/DetailView/DetailView2.php');
+
+class DetailDraftView extends DetailView2
+{
+
+    public function setup(
```
Comment (daniel-samson) [inline modules/Emails/include/DetailDraftView/DetailDraftView.php]: added phpdoc

File: modules/Emails/views/view.savedraftemail.php (hunk 3 above)
Comment (daniel-samson) [inline modules/Emails/views/view.savedraftemail.php]: fixed

File: modules/Emails/views/view.savedraftemail.php (hunk 6 above)
Comment (daniel-samson) [inline modules/Emails/views/view.savedraftemail.php]: fixed

File: modules/Emails/views/view.savedraftemail.php (hunk 4 above)
Comment (daniel-samson) [inline modules/Emails/views/view.savedraftemail.php]: updated phpdocs

File: modules/Emails/views/view.sendemail.php (hunk 5 above)
Comment (daniel-samson) [inline modules/Emails/views/view.sendemail.php]: updated phpdocs

**Pull request merged**